File: BeerSample/BeerSample/BeerViewController.m
//
//  ViewController.m
//  BeerSample
//
//  Created by Vinita Miranda on 3/6/15.
//  Copyright (c) 2015 Vinita Miranda. All rights reserved.
//

#import "BeerViewController.h"
#import "BeerTableViewCell.h"
#import "ProductViewController.h"
@interface BeerViewController ()
@property(nonatomic,weak)IBOutlet UITableView *tableView;
@property(nonatomic)NSArray *beerArray;
@end

@implementation BeerViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.navigationController.navigationBar.translucent = NO;

    [self fetchBeerData];
}

-(void)viewWillAppear:(BOOL)animated
{
    self.title = @"Beer";
}

-(void)fetchBeerData
{
    NSURLSession *session = [NSURLSession sharedSession];
    NSURLSessionDataTask *task = [session dataTaskWithURL:[NSURL URLWithString:@"http://ontariobeerapi.ca:80/beers/"] completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        
        self.beerArray = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:nil];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.tableView reloadData];

        });
        
        
    }];
    [task resume];
    
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


- (NSInteger)tableView:(UITableView *)tableView
 numberOfRowsInSection:(NSInteger)section
{
    return [self.beerArray count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    BeerTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"BeerTableViewCell"];
    NSDictionary *beerDictionary = [self.beerArray objectAtIndex:indexPath.row];
    cell.titleLabel.text = [beerDictionary objectForKey:@"name"];
    cell.typeLabel.text = [beerDictionary objectForKey:@"type"];
    cell.ratingLabel.text = [beerDictionary objectForKey:@"abv"];

    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        
        NSData *data = [NSData dataWithContentsOfURL:[NSURL URLWithString:[beerDictionary objectForKey:@"image_url"]]];
        
        if(data)
        {
            UIImage *beerImage = [UIImage imageWithData:data];
            if(beerImage)
            {
                dispatch_async(dispatch_get_main_queue(), ^{
                    UIImage *beerImage = [UIImage imageWithData:data];
                    cell.iconImageView.image = beerImage;
                });
            }
        }
        
    });
    
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 80.0;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    self.title = @"";
    ProductViewController *productViewController = [self.storyboard instantiateViewControllerWithIdentifier:@"ProductViewController"];
    productViewController.beerDictionary = [self.beerArray objectAtIndex:indexPath.row];
    [self.navigationController pushViewController:productViewController animated:YES];

}



@end

File: Swift/BeerSampleSwift/BeerSampleSwift/BeerViewController.swift
//
//  BeerViewController.swift
//  BeerSampleSwift
//
//  Created by Vinita Miranda on 5/25/15.
//  Copyright (c) 2015 Vinita Miranda. All rights reserved.
//

import UIKit

class BeerViewController: UIViewController,DataManagerProtocol,UITableViewDelegate,UITableViewDataSource    
{
    var beerArray = NSMutableArray()

    @IBOutlet weak var tableView: UITableView!
    override func viewDidLoad() {
        super.viewDidLoad()
        self.tableView.registerNib(UINib(nibName: "CustomTableViewCell", bundle: NSBundle.mainBundle()), forCellReuseIdentifier: "CustomTableViewCell")
        self.fetchBeerData()
        self.navigationController!.navigationBar.translucent = false;

    }
    
    override func viewWillAppear(animated: Bool)
    {
        self.title = "Beer"
    }

    func fetchBeerData()
    {
        var dataManager : DataManager = DataManager()
        dataManager.delegate = self
        dataManager.fetchDataFromUrl("http://ontariobeerapi.ca:80/beers/")
    }

    
    func datafetchCompleted(result : NSMutableArray)
    {
        self.beerArray = result
        dispatch_async(dispatch_get_main_queue())
        {
            self.tableView.reloadData()
        }
    }
    
    func datafetchFailed(error : NSError)
    {
        
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.beerArray.count
    }
    
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell
    {
        var cell : CustomTableViewCell = tableView.dequeueReusableCellWithIdentifier("CustomTableViewCell") as! CustomTableViewCell
        var beerDictionary: NSDictionary = NSDictionary()
        beerDictionary  = self.beerArray[indexPath.row] as! NSDictionary
        cell.titleLabel.text = beerDictionary["name"] as? String
        cell.typeLabel.text = beerDictionary["type"] as? String
        cell.ratingLabel.text = beerDictionary["abv"] as? String
        
        var imageName : String? =  beerDictionary["image_url"] as? String

        ImageLoader.sharedLoader.imageForUrl(imageName!, completionHandler:{(image: UIImage?, url: String) in
            cell.iconImageView.image = image
        })
        
        return cell
    }
    
    
    func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        return 80
    }
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath)
    {
        var destinationVC : ProductViewController = (self.storyboard?.instantiateViewControllerWithIdentifier("ProductViewController") as? ProductViewController)!
        destinationVC.beerDictionary = self.beerArray[indexPath.row] as? NSDictionary
        self.navigationController!.pushViewController(destinationVC, animated: true)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
}
